File: 2022/02/28/countBits.ts
/**
 * https://leetcode.com/problems/counting-bits/
 * @author Andrew Jarombek
 * @since 2/28/2022
 */

function countBits(n: number): number[] {
    let result = [];

    for (let i = 0; i <= n; i++) {
        let bits = 0;
        let value = i;

        while (value > 0) {
            bits += value % 2;
            value >>= 1;
        }

        result.push(bits);
    }

    return result;
}

File: 2022/07/14/java/CountBits.java
/**
 * https://leetcode.com/problems/counting-bits/
 * @author Andrew Jarombek
 * @since 7/14/2022
 */

public class CountBits {
    public static int[] countBits(int n) {
        int[] result = new int[n+1];

        for (int i = 0; i <= n; i++) {
            int ones = 0;
            int num = i;

            while (num > 0) {
                ones += num & 1;
                num >>= 1;
            }

            result[i] = ones;
        }

        return result;
    }

    public static void main(String[] args) {
        var result = countBits(4);

        assert result[0] == 0;
        assert result[1] == 1;
        assert result[2] == 1;
        assert result[3] == 2;
        assert result[4] == 1;
    }
}
